### DISCUSSION TITLE
Coding Questions

### DISCUSSION THREAD

**Discussion by FallowLand** [bitcraze/discussions#199] [Q&A] [answered]
Hi Bitcraze,

I have some questions about coding:
a. Why are some codes written in python and some in c++? I noticed that to change the LED color, I have to use c++. Whereas, to fly the following flight path (https://www.bitcraze.io/documentation/tutorials/getting-started-with-stem-drone-bundle/) I use python
b. If I wanted to do the same flight in the previous link but using c++, do I create a new .h file or do I put the code in a existing .h file?
c. Generally speaking, is it better to do developement in c++ (Virtual Machine) or python (Windows Machine)?

**Comment by krichardsson** [comment, marked as answer]
Hi!

> a. Why are some codes written in python and some in c++? I noticed that to change the LED color, I have to use c++. Whereas, to fly the following flight path (https://www.bitcraze.io/documentation/tutorials/getting-started-with-stem-drone-bundle/) I use python

Code that is running in a PC is written in python to make it as platform independent as possible, while code that is running in the Crazyflie is written in C (not C++) since it is a better language for low level control in embedded systems. The [python lib](https://github.com/bitcraze/crazyflie-lib-python) and [client](https://github.com/bitcraze/crazyflie-clients-python) are in python while the [Crazyflie firmware](https://github.com/bitcraze/crazyflie-firmware) is in C.

> b. If I wanted to do the same flight in the previous link but using c++, do I create a new .h file or do I put the code in a existing .h file?

If you want to run the code in the Crazyflie, you should write it as an Crazyflie application (in C). There is some information about [development basics](https://www.bitcraze.io/documentation/tutorials/getting-started-with-development/) and [the app layer](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/app_layer/). If you want to write C++ code that runs in the PC, you should probably use the driver from [Crazyswarm](https://crazyswarm.readthedocs.io/en/latest/index.html)

> c. Generally speaking, is it better to do developement in c++ (Virtual Machine) or python (Windows Machine)?

It depends on what you want to do. If you want to fly nice trajectories, python in the PC is by far the easiest solution. For low level hardware integration, C and the firmware is the way to go.
